src/appleseed/renderer/kernel/rendering/final/adaptivetilerenderer.cpp

```diff
@@ -357,66 +354,64 @@ namespace
                     abort_switch,
                     batch_size,
                     framebuffer,
-                    second_framebuffer,
+                    second_framebuffer.get(),
                     tile_origin_x,
                     tile_origin_y,
                     frame,
                     frame_width,
                     pass_hash,
                     aov_count);
 
-                // Check if this was the last batch.
-                if (remaining_samples - batch_size < 1)
+                const AABB2u& block_image_bb = AABB2i::intersect(framebuffer->get_crop_window(), pb.m_surface);
+
+                // Evaluate block's noise amount.
+                pb.m_block_error = FilteredTile::compute_tile_variance(
+                    block_image_bb,
+                    framebuffer,
+                    second_framebuffer.get());
+
+                if (batch_size < m_params.m_batch_size)
                 {
+                    // It was the last batch.
                     finished_blocks.push_back(pb);
                 }
-                else
+                else if (pb.m_block_error <= m_params.m_noise_threshold)
                 {
-                    const AABB2u& block_image_bb = AABB2i::intersect(framebuffer->get_crop_window(), pb.m_surface);
-
-                    // Evaluate block's noise amount.
-                    pb.m_block_error = FilteredTile::compute_tile_variance(
-                        block_image_bb,
-                        framebuffer,
-                        second_framebuffer);
-
-                    // Decide if the blocks needs to be splitted, sampled or if it has converged.
-                    if (pb.m_block_error <= m_params.m_noise_threshold)
+                    // The block has converged.
+                    pb.m_converged = true;
+                    finished_blocks.push_back(pb);
+                }
+                else if (pb.m_block_error <= m_params.m_splitting_threshold)
+                {
+                    // The block needs to be splitted.
+                    if (pb.m_main_axis == PixelBlock::Axis::Horizontal
+                        && block_image_bb.extent(0) >= BlockSplittingThreshold)
                     {
-                        pb.m_converged = true;
-                        finished_blocks.push_back(pb);
+                        split_pixel_block(
+                            pb,
+                            rendering_blocks,
+                            static_cast<int>(block_image_bb.min.x)
```

Comment:
What is the `- 0.5f` part for?

If all you want is the middle of the box, you can use `block_image_bb.center(0)` (but please make sure it does exactly what you need).

Reply:
I don't remember why I did it this way. I think it's because I wanted to truncate the value

Reply:
Maybe you mean "round the value" instead of "truncate the value"? Because that's what happens when you truncate `x + 0.5f`: you round `x` to the nearest integer.

In any case that code should be checked and clarified.

Reply:
There isn't any `center` for integral bboxes, which make sense